shopinvader_sale_coupon/services/cart.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2021 Camptocamp SA
+# @author Iván Todorovich <ivan.todorovich@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+from odoo.exceptions import UserError
+
+from odoo.addons.component.core import Component
+
+
+class CartService(Component):
+    _inherit = "shopinvader.cart.service"
+
+    # Public Methods
+
+    def apply_coupon(self, code):
+        cart = self._get()
+        self._apply_coupon(cart, code)
+        return self._to_json(cart)
+
+    # Private Methods
+
+    def _apply_coupon(self, cart, code):
+        """
+        Apply a coupon or promotion code.
+        :return: Boolean
+        """
+        SaleCouponApplyCode = self.env["sale.coupon.apply.code"].sudo()
+        status = SaleCouponApplyCode.apply_coupon(cart, code)
+        if status.get("error"):
+            raise UserError(status["error"])
```

Comment:
what about `not_found`? I think it's better to simply return the result and let the frontend handle the message

Reply:
I asked myself the same and then ended up aligning to what odoo does in `website_sale_coupon`:

https://github.com/odoo/odoo/blob/a61d459a200a56fba74f655f73b10d1ddd8e3a1e/addons/website_sale_coupon/controllers/main.py#L9-L17

I'll change it